lua/ttt2/libraries/spawn.lua

```diff
@@ -0,0 +1,290 @@
+	local spawnPoints = spawn.GetSpawnPointsAroundSpawn(unsafePos)
+
+	for i = 1, #spawnPoints do
+		local spawnPoint = spawnPoints[i]
+
+		if not spawn.IsSpawnPointSafe(spawnPoint, false) then continue end
+
+		return spawnPoint
+	end
+end
+
+---
+-- Returns a list of all spawn entities
+-- @param boolean shouldShuffle whether the table should be shuffled
+-- @param boolean forceAll used unless absolutely necessary (includes info_player_start spawns)
+-- @return table Returns a table of unsafe spawn entities
+-- @realm server
+function spawn.GetPlayerSpawnEntities(shouldShuffle, forceAll)
+	local tbl = {}
+
+	for i = 1, #spawnTypes do
+		local classname = spawnTypes[i]
+		local entsTbl = entsFindByClass(classname)
+
+		for k = 1, #entsTbl do
+			if entsTbl[k].markAsRemoved then continue end
+
+			tbl[#tbl + 1] = entsTbl[k]
+		end
+	end
+
+	-- Don't use info_player_start unless absolutely necessary, because eg. TF2
+	-- uses it for observer starts that are in places where players cannot really
+	-- spawn well. At all.
+	if forceAll or #tbl == 0 then
+		local startTbl = entsFindByClass("info_player_start")
+
+		for k = 1, #startTbl do
+			if startTbl[k].markAsRemoved then continue end
+
+			tbl[#tbl + 1] = startTbl[k]
+		end
+	end
+
+	if shouldShuffle then
+		table.Shuffle(tbl)
+	end
+
+	return tbl
+end
+
+---
+-- Gets a table of spawnpoints on the map.
+-- @return table Returns a table of spawnpoints as @{Vector}s
+-- @realm server
+function spawn.GetPlayerSpawnPointTable()
+	local spawnEnts = spawn.GetPlayerSpawnEntities(true, false)
+	local spawnPoints = {}
+
+	for i = 1, #spawnEnts do
+		spawnPoints[i] = spawnEnts[i]:GetPos()
+	end
+
+	return spawnPoints
+end
+
+---
+-- Gets a safe random player spawn entity. If no free and safe spawnpoint is found, it tries
+-- to create its own by searching around existing ones.
+-- @param Player ply The player that should receive their spawn point
+-- @return Entity A safe spawn entity
+-- @realm server
+function spawn.GetRandomPlayerSpawnEntity(ply)
+	-- One might think that we have to regenerate our spawnpoint
+	-- cache. Otherwise, any riggedSpawnPoints spawn entities would not get reused, and
+	-- MORE new entities would be made instead. In reality, the map cleanup at
+	-- round start will remove our riggedSpawnPoints spawns, and we'll have to create new
+	-- ones anyway.
+	if #spawn.cachedSpawnEntities == 0 or not IsTableOfEntitiesValid(spawn.cachedSpawnEntities) then
+		spawn.cachedSpawnEntities = spawn.GetPlayerSpawnEntities(true, false)
+	end
+
+	if #spawn.cachedSpawnEntities == 0 then
+		Error("No spawn entity found!\n")
+
+		return
+	end
+
+	-- the table should be shuffled for each spawn point calculation to improve randomness
+	table.Shuffle(spawn.cachedSpawnEntities)
+
+	-- Optimistic attempt: assume there are sufficient spawns for all and one is free
+	for i = 1, #spawn.cachedSpawnEntities do
+		local spawnEntity = spawn.cachedSpawnEntities[i]
+
+		if not IsValid(ply) or not ply:IsTerror() then
```

Comment:
Why are these checks now needed?

Reply:
These came from original TTT. They are needed because as a spectator, your spawnpoint should always be valid.